src/Network/Wai/Middleware/Auth/OAuth2.hs

```diff
@@ -104,46 +87,58 @@ instance AuthProvider OAuth2 where
-        let redirectUrl =
-              getRedirectURI $
-              appendQueryParams
-                (OA2.authorizationUrl oauth2)
-                (maybe [] ((: []) . ("scope", )) scope)
-        return $
-          responseLBS
-            status303
-            [("Location", redirectUrl)]
-            "Redirect to OAuth2 Authentication server"
-      ["complete"] ->
-        let params = queryString req
-        in case lookup "code" params of
-             Just (Just code) -> do
-               man <- getGlobalManager
-               eRes <- OA2.fetchAccessToken man oauth2 $ getExchangeToken code
-               case eRes of
-                 Left err    -> onFailure status501 $ S8.pack $ show err
-                 Right token -> onSuccess $ encodeToken token
-             _ ->
-               case lookup "error" params of
-                 (Just (Just "access_denied")) ->
-                   onFailure
-                     status403
-                     "User rejected access to the application."
-                 (Just (Just error_code)) ->
-                   onFailure status501 $ "Received an error: " <> error_code
-                 (Just Nothing) ->
-                   onFailure status501 $
-                   "Unknown error connecting to " <>
-                   encodeUtf8 (getProviderName oa2)
-                 Nothing ->
-                   onFailure
-                     status404
-                     "Page not found. Please continue with login."
-      _ -> onFailure status404 "Page not found. Please continue with login."
-
+    man <- getGlobalManager
+    oauth2Login
+      oauth2
+      man
+      oa2Scope
+      (getProviderName oa2)
+      req
+      suffix
+      onSuccess
+      onFailure
+  refreshLoginState OAuth2 {..} req user = do
+    authEndpointURI <- parseAbsoluteURI oa2AuthorizeEndpoint
+    accessTokenEndpointURI <- parseAbsoluteURI oa2AccessTokenEndpoint
+    let oauth2 =
+          OA2.OAuth2
+          { oauthClientId = getClientId oa2ClientId
+          , oauthClientSecret = getClientSecret oa2ClientSecret
+          , oauthOAuthorizeEndpoint = authEndpointURI
+          , oauthAccessTokenEndpoint = accessTokenEndpointURI
+          -- Setting callback endpoint to `Nothing` below is a lie.
+          -- We do have a callback endpoint but in this context
+          -- don't have access to the function that can render it.
+          -- We get away with this because the callback endpoint is
+          -- not needed for obtaining a refresh token, the only
+          -- way we use the config here constructed.
+          , oauthCallback = Nothing
+          }
+    man <- getGlobalManager
+    let loginState = authLoginState user
+    case decodeToken loginState of
+      Left _ -> pure Nothing
+      Right tokens -> do
+        CTime now <- epochTime
+        if tokenExpired user now tokens then do
+          rRes <- refreshTokens tokens man oauth2
```

Comment:
It is better to move the lines above here since neither `oauth2` nor `man` are needed in the `else` case:

```haskell
    authEndpointURI <- parseAbsoluteURI oa2AuthorizeEndpoint
    accessTokenEndpointURI <- parseAbsoluteURI oa2AccessTokenEndpoint
    let oauth2 =
          OA2.OAuth2
          { oauthClientId = getClientId oa2ClientId
          , oauthClientSecret = getClientSecret oa2ClientSecret
          , oauthOAuthorizeEndpoint = authEndpointURI
          , oauthAccessTokenEndpoint = accessTokenEndpointURI
          -- Setting callback endpoint to `Nothing` below is a lie.
          -- We do have a callback endpoint but in this context
          -- don't have access to the function that can render it.
          -- We get away with this because the callback endpoint is
          -- not needed for obtaining a refresh token, the only
          -- way we use the config here constructed.
          , oauthCallback = Nothing
          }
    man <- getGlobalManager
```

Reply:
Makes sense, thanks!